File: 350. Intersection of Two Arrays II.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> hash;
        for(auto i: nums1)
            hash[i]++;
        vector<int> res;
        for(auto i: nums2){
            if(hash[i]>0){
                res.push_back(i);
                hash[i]--;
            }
        }
        return res;
    }
};

class Solution2 {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        int N1 = nums1.size();
		int N2 = nums2.size();
		sort(nums1.begin(), nums1.end());
		sort(nums2.begin(), nums2.end());
		vector<int> result;
		int i = 0, j = 0;
		while (i < N1 && j < N2) {
			if (nums1[i] < nums2[j]) ++i;
			else if (nums2[j] < nums1[i]) ++j;
			else {
				result.push_back(nums1[i]);
				++i;
				++j;
			}
		}
		return result;
    }
};

File: 350. Intersection of Two Arrays II.py
class Solution(object):

    def intersect(self, nums1, nums2):

        """

        :type nums1: List[int]

        :type nums2: List[int]

        :rtype: List[int]

        """

        hash = collections.Counter()

        for i in nums1:

            hash[i]+=1

        res = []

        for j in nums2:

            if hash[j]>0:

                hash[j]-=1

                res.append(j)

        return res
